language/move-prover/tests/sources/functional/macro_verification.move

```diff
@@ -0,0 +1,100 @@
+// flag: --trace
+
+// CVC5 does currently not terminate on this:
+// exclude_for: cvc5
+
+/// This file contains some simulated expansions of functional macros and sketches how they can be verified.
+module 0x42::FunMacros {
+    use Std::Vector;
+
+    /// Simulates `foreach!(v, |x| *x = *x + 1 spec  { ensures x == old(x) + 1; })`
+    /// where `macro foreach!<T>(v: &mut vector<T>, action: |&mut T|)`
+    ///
+    /// For the foreach macro the loop invariant can be synthesized from a post condition associated
+    /// with the lambda.
+    fun foreach(v: &mut vector<u64>) {
+        let i = 0;
+        while (i < Vector::length(v)) {
+            let x = Vector::borrow_mut(v, i);
+            *x = *x + 1;
+            i = i + 1;
+        }
+        spec {
+            invariant i >= 0 && i <= len(v);
+            invariant len(v) == len(old(v));
+            invariant forall j in 0..i: v[j] == old(v)[j] + 1; // lambda substituted
+            invariant forall j in i..len(v): v[j] == old(v)[j];
+        };
+    }
+    spec foreach {
+        ensures len(v) == len(old(v));
+        ensures forall i in range(v): v[i] == old(v)[i] + 1; // succeeds
+        ensures forall i in range(v): v[i] == old(v)[i] + 2; // fails
+    }
+
+    /// Simulates `reduce!(v, 0, |sum, x| *sum = *sum + *x ) spec { invariant sum == old(sum) + x; })`
+    /// where `macro reduce!<T, R>(v: &vector<T>, neutral: R, reducer: |&mut R, &T|): R`.
+    ///
+    /// Because the elements of the vector are combined via the reducer, we cannot specify this with
+    /// a quantifier, however, we can use a recursive helper function.
+    fun reduce(v: &vector<u64>) : u64 {
+        let i = 0;
+        let sum = 0;
+        while (i < Vector::length(v)) {
+            let x = Vector::borrow(v, i);
+            sum = sum + *x;
+            i = i + 1;
+        }
+        spec {
+            invariant i >= 0 && i <= len(v);
+            invariant sum == spec_sum(v, i);
+        };
+        sum
+    }
+    spec reduce {
+        ensures result == spec_sum(v, len(v));     // succeeds
+        // Below we use a premise for the failure case to constraint model size.
+        ensures len(v) <= 4 ==> result == spec_sum(v, len(v)) + 1; // fails
+    }
+    spec fun spec_sum(v: vector<u64>, end: num): num {
```

Comment:
I have a question/comment unrelated to this PR.  Suppose I wanted to prove a property of the function spec_sum, for example, that the result of spec_sum(x) >= 0 for all x.  Is it possible to do this proof today in Move Prover?

Reply:
This is currently not possible -- spec funs cannot have pre/post conditions.

Reply:
Maybe this does not require postconditions?  I think it's just proving a condition that doesn't have an associated state change (which is something I might want for other reasons).  Please tell me if I'm interpreting the question correctly.

I.e., we're trying to prove this:
forall x:u64 spec_sum(x) >= 0

A kludgey way to do it might be to have an identity function:
fun foo(v: vector<u64>): vector<u64> {
     v
}
spec foo {
      ensures forall x:u64 spec_sum(x) >= 0;
}

Obviously, we could add the necessary syntax to prove "lemmas" like this without the kludge of an identity function.